### PR TITLE
Fixed wrong API permission for GreenField/StorePaymentMethodsController.cs

### PR DESCRIPTION
**Pull request by woutersamaey**
Viewing store payment methods needs CanViewStoreSettings instead of CanModifyStoreSettings as per the docs

### CODE REVIEW

**Comment by woutersamaey** [comment]
@Kukks and @ndeet this may interest you. The permissions are wrong in code, but correct in the docs.

**Review by Kukks — APPROVED** [review]
i think it is an actual bug but maybe @NicolasDorier has some more insight?

**Review by Kukks — CHANGES_REQUESTED** [review]
So, potentially this is on purpose because for `GetLightningNetworkPaymentMethod`, it is also `CanModifyStoreSettings`.

Most likely, the reasoning is that the payment method contains sensitive information that would compromise the security of your payment method, even if it is only with viewing rights. For example, the lightning connection string could contain the access token/macaroon to control and spend from your lightning node.

**Comment by woutersamaey** [comment]
@Kukks this cuts both ways, doesn't it?

Either you compromise on security by weakening your LN config security
Or, you weaken your installation by creating and storing API keys in an eCommerce system that can be hacked and more control can be gained than was needed.

The best solution would be to censor the potentially dangerous fields if the lesser permissions is used, and include the sensitive info when an API key with more permissions is used.

That way you get security for both cases, but I guess some refactoring would be involved? Or can you simply filter out the values as a filter before data is output from the controller? (i.e. set the dangerous values to NULL or something like `*HIDDEN*` or `***`

**Comment by Kukks** [comment]
> @Kukks this cuts both ways, doesn't it?
> 
> Either you compromise on security by weakening your LN config security
> Or, you weaken your installation by creating and storing API keys in an eCommerce system that can be hacked and more control can be gained than was needed because the API key was created with otherwise less permissions.
> 
> The best solution would be to censor the potentially dangerous fields if the lesser permissions is used, and include the sensitive info when an API key with more permissions is used.
> 
> That way you get security for both cases, but I guess some refactoring would be involved? Or can you simply filter the values before data is output from the controller? (i.e. set the dangerous values to something like `*HIDDEN*` or `***`
> 
> Just trying to find an easy solution here...

Yeah, this is probably the best way.

**Comment by Kukks** [comment]
Fixed up the PR with the discussed changes.

**Comment by woutersamaey** [comment]
Probably nitpicking here, but I believe `*HIDDEN*` would've been less confusing to newbies over that `###` you used.
I fear this will lead to more support questions.

Maybe better yet would be something like `*NEED CanModifyStoreSettings PERMISSION TO VIEW*`

File: BTCPayServer.Tests/GreenfieldAPITests.cs [hunk 1]
```diff
@@ -1986,6 +1987,29 @@ void VerifyOnChain(Dictionary<string, GenericPaymentMethodData> dictionary)
...
+            Assert.True(methods.TryGetValue(new PaymentMethodId("BTC", PaymentTypes.LightningLike).ToStringNormalized(), out var item));
+            var lightningNetworkPaymentMethodBaseData =Assert.IsType<JObject>(item.Data).ToObject<LightningNetworkPaymentMethodBaseData>();
+            Assert.Equal("###", lightningNetworkPaymentMethodBaseData.ConnectionString);
+            
+            
+            methods = await adminClient.GetStorePaymentMethods(store.Id);
+            
+            Assert.True(methods.TryGetValue(new PaymentMethodId("BTC", PaymentTypes.LightningLike).ToStringNormalized(), out item));
+            lightningNetworkPaymentMethodBaseData =Assert.IsType<JObject>(item.Data).ToObject<LightningNetworkPaymentMethodBaseData>();
+            Assert.NotEqual("###", lightningNetworkPaymentMethodBaseData.ConnectionString);
```
Comment (Kukks) [inline BTCPayServer.Tests/GreenfieldAPITests.cs:2011]:
```suggestion
            Assert.NotEqual("*NEED CanModifyStoreSettings PERMISSION TO VIEW*", lightningNetworkPaymentMethodBaseData.ConnectionString);
```

File: BTCPayServer.Tests/GreenfieldAPITests.cs [hunk 2]
```diff
@@ -1986,6 +1987,29 @@ void VerifyOnChain(Dictionary<string, GenericPaymentMethodData> dictionary)
...
+            
+            
+           await  adminClient.UpdateStoreLightningNetworkPaymentMethod(store.Id, "BTC",
+                new UpdateLightningNetworkPaymentMethodRequest(
+                    tester.GetLightningConnectionString(LightningConnectionType.CLightning, true), true));
+            methods = await viewerOnlyClient.GetStorePaymentMethods(store.Id);
+            
+            Assert.True(methods.TryGetValue(new PaymentMethodId("BTC", PaymentTypes.LightningLike).ToStringNormalized(), out var item));
+            var lightningNetworkPaymentMethodBaseData =Assert.IsType<JObject>(item.Data).ToObject<LightningNetworkPaymentMethodBaseData>();
+            Assert.Equal("###", lightningNetworkPaymentMethodBaseData.ConnectionString);
```
Comment (Kukks) [inline BTCPayServer.Tests/GreenfieldAPITests.cs:2004]:
```suggestion
            Assert.Equal("*NEED CanModifyStoreSettings PERMISSION TO VIEW*", lightningNetworkPaymentMethodBaseData.ConnectionString);
```

File: BTCPayServer/Payments/PaymentTypes.Lightning.cs [hunk 3]
```diff
@@ -65,12 +65,20 @@ public override string GetTransactionLink(BTCPayNetworkBase network, string txId
...
                 {
-                    ConnectionString = lightningSupportedPaymentMethod.GetDisplayableConnectionString()
+                    ConnectionString = lightningSupportedPaymentMethod.IsInternalNode
+                        ?
+                        lightningSupportedPaymentMethod.GetDisplayableConnectionString()
+                        :
+                        canModifyStore
+                            ? lightningSupportedPaymentMethod.GetDisplayableConnectionString()
+                            :
+                            "###"
```
Comment (Kukks) [inline BTCPayServer/Payments/PaymentTypes.Lightning.cs:81]:
```suggestion
                            "*NEED CanModifyStoreSettings PERMISSION TO VIEW*"
```

**Review by Kukks — CHANGES_REQUESTED** [review]

**Comment by woutersamaey** [comment]
Awesome!

**Pull request merged**